ISO-8859-1 as default for application/x-www-form-urlencoded

candrews opened the pull request:
The IANA registration at https://www.w3.org/TR/2016/WD-html51-20160310/iana.html#application-x-www-form-urlencoded specifies that a US-ASCII encoding should be used by default. In practice, it's also what browsers and servers expect when no charset parameter is specified.

When writing with content type application/x-www-form-urlencoded, if the charset is the default, leave the charset off of the content-type. Since it's the default, it need not be specified, and some (particularly) servers don't work correctly when a charset is specified on application/x-www-form-urlencoded. See https://bugzilla.mozilla.org/show_bug.cgi?id=241540 for background.

Fixes https://github.com/spring-projects/spring-framework/issues/22548

candrews committed:
ISO-8859-1 as default for application/x-www-form-urlencoded

The IANA registration at https://www.w3.org/TR/2016/WD-html51-20160310/iana.html#application-x-www-form-urlencoded specifies that a US-ASCII charset should be used by default. In practice, it's also what browsers and servers expect when no charset parameter is specified.

When writing with content type application/x-www-form-urlencoded, if the charset is the default, leave the charset off of the content-type. Since it's the default, it need not be specified, and some (particularly) servers don't work correctly when a charset is specified on application/x-www-form-urlencoded. See https://bugzilla.mozilla.org/show_bug.cgi?id=241540 for background.

andrei-ivanov commented on spring-web/src/main/java/org/springframework/http/converter/FormHttpMessageConverter.java:
```diff
@@ -312,14 +312,24 @@ private void writeForm(MultiValueMap<String, Object> formData, @Nullable MediaTy
 		}
+		
+		if (MediaType.APPLICATION_FORM_URLENCODED.includes(contentType) && charset.equals(DEFAULT_APPLICATION_FORM_URLENCODED_CHARSET) ) {
+			contentType = (MediaType.APPLICATION_FORM_URLENCODED;	
```
does this line compile?

XYUU commented:
This change will result in many historical projects not working after the update, please be careful to merge.

candrews commented:
> This change will result in many historical projects not working after the update, please be careful to merge.

I'm not so sure that's true.

This form data is supposed to be url encoded to only US ASCII, and that's what browsers seem to do. If the form does something else, it should be setting the charset (ex, if it's sending url encoded data but not encoding utf-8 characters, it should set the charset to utf-8, and at least in the case of Spring, that is what it does).

This change shouldn't break anything, brings Spring in line with what is done in practice and what is specified by the rfc, and fixes currently broken use cases (such as where Spring is used to make form data for a server that doesn't accept the charset parameter, which according to the iana registration for this content type, not accepting a charset parameter is a valid behavior, which is something Mozilla discovered over a decade ago which can be seen from the bug report I linked to).

If you believe this would break something, can you please more information?

snicoll commented:
@candrews in the future there is no need to open an issue and PR now we've moved the issue tracker to Github, one is enough. Also, I'd expect such a change to contain unit tests to exercice the behaviour.

XYUU commented:
For programs that do not currently have a specified character set, UTF-8 will be used to decode and encode. If it is currently working, it will not work after this merge. The problem caused by this historical reason has brought other problems, so I think it should be carefully considered.

XYUU commented:
Moreover, the code of your change is not elegant, and it is not uniform with webflux. For example, [FormHttpMessageReader ](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/codec/FormHttpMessageReader.java)and [FormHttpMessageWriter](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/codec/FormHttpMessageWriter.java) still use UTF-8.

candrews commented:
> I'd expect such a change to contain unit tests to exercice the behaviour.

Point taken - I've updated this PR with tests proving the behavior.

> Moreover, the code of your change is not elegant, 

Feedback welcome! If you have something specific, please let me know and I'll do my best to address it.

> and it is not uniform with webflux. For example, [FormHttpMessageReader ](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/codec/FormHttpMessageReader.java)and [FormHttpMessageWriter](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/codec/FormHttpMessageWriter.java) still use UTF-8.

I'm not a webflux user at the moment so I didn't notice those classes. I've updated this PR with corresponding changes to those as well - thank you for pointing it out.

rstoyanchev commented:
See my response under #22548.
